feat: show dot for unread messages

marhaasa opened the pull request:
Added a small change to display a red dot behind channels that have unread messages since application start

marhaasa committed:
feat: show red dot for unread messages

Added a small change to display a red dot behind channels that have
unread messages since application start

marhaasa committed:
fix: switch to ningen@v3 and arikawa API

Reworked the implementation of showing channels with unread messages and
mentions to use ningen and arikawa API as suggested. The benefit of this
implementation is that reading unread messages and mentions now updates
the state for discord as a whole instead of just inside the discordo.

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,124 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+		app.QueueUpdateDraw(func() {
+			gt.refreshChannelDisplay(channelID)
+		})
+	}()
```
Listen to the `Update` handler provided by ningen instead of delay.
```go
// state.go
discordState.AddHandler(func(event *ningen.UpdateEvent) {
// ...
})
```

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -192,6 +290,13 @@ func (gt *guildsTree) onSelected(node *tview.TreeNode) {
 		}
 
+		// Mark channel as read when selected
+		gt.markChannelAsRead(channel.ID)
```
https://pkg.go.dev/github.com/diamondburned/ningen/v3@v3.0.0/states/read#State.MarkRead

ayn2op commented on cmd/state.go:
```diff
@@ -111,6 +111,11 @@ func onMessageCreate(msg *gateway.MessageCreateEvent) {
+		if msg.Author.ID != discordState.Ready().User.ID {
+			app.guildsTree.refreshChannelDisplay(msg.ChannelID)
+			app.Draw()
+		}
```
https://pkg.go.dev/github.com/diamondburned/ningen/v3@v3.0.0/states/read#State.MarkUnread

marhaasa committed:
Fix ningen MarkUnread and UpdateEvent

Based on the comments I have now moved to use ningen's MarkUnread and
UpdateEvent to move away from janky delays and hacks.

marhaasa committed:
Add unread navigation

Added the ability to move to next and prev unread message channel.

marhaasa committed:
Remove select upon navigation

Now the navigation only moves to the next or prev unread, but does not
select it in case the user wants to keep it unread. Also included direct
messages to the unread navigation.

ayn2op commented on internal/config/keys.go:
```diff
@@ -30,6 +30,9 @@ type (
 		CollapseParentNode string `toml:"collapse_parent_node"`
 		MoveToParentNode   string `toml:"move_to_parent_node"`
+
+		NextUnread     string `toml:"next_unread"`
+		PreviousUnread string `toml:"previous_unread"`
```
These keymaps are not working in order.

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+func (gt *guildsTree) markChannelAsRead(channelID discord.ChannelID) {
+	if discordState == nil {
+		return
+	}
```
Redundant check: at this point, `discordState` is guaranteed to be non-nil.

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+func (gt *guildsTree) getChannelDisplayName(channel discord.Channel) string {
+	name := gt.channelToString(channel)
+	
+	if discordState != nil {
```
Redundant

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -273,8 +525,6 @@ func (gt *guildsTree) yankID() {
 	}
 
-	// Reference of a tree node in the guilds tree is its ID.
-	// discord.Snowflake (discord.GuildID and discord.ChannelID) have the String method.
```
Why?

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+	}
+	
+	latestMsgID := msgs[0].ID
+	go gt.ackChannel(channelID, latestMsgID)
```
Use `discordState.ReadState.MarkRead` instead: https://pkg.go.dev/github.com/diamondburned/ningen/v3@v3.0.0/states/read#State.MarkRead

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+		return
+	}
+	
+	latestMsgID := msgs[0].ID
```
https://pkg.go.dev/github.com/diamondburned/ningen/v3@v3.0.0#State.LastMessage

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+		indication := discordState.ChannelIsUnread(channel.ID, opts)
+		switch indication {
+		case ningen.ChannelMentioned:
+			return fmt.Sprintf("%s @", name)
```
`name + " @"` and `name + " ●"`

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+	}
+	
+	slog.Debug("marked channel as read", "channel_id", channelID, "message_id", messageID)
+}
```
`MarkRead` sends the `ack` request for you: https://pkg.go.dev/github.com/diamondburned/ningen/v3@v3.0.0/states/read#State.MarkRead

marhaasa replied:
I am having trouble getting the MarkRead to actually mark channels as read. Yes - it works in the current discordo session, but it does not work cross platform.

marhaasa replied:
If this is wanted I'll let it be, but for me personally I would like that reading channels in Discordo marks them as read cross platform.

ayn2op replied:
What do you mean "cross-platform"?

marhaasa replied:
That the messages are marked as read by Discord's servers - not just discordo session. The implementation I have presented actually marks channels as read when starting a new discordo session, viewing on mobile app and on web.

marhaasa replied:
I can't seem to get the same functionality with ningen State.MarkRead.

ayn2op commented on cmd/guilds_tree.go:
```diff
@@ -70,32 +73,267 @@ func (gt *guildsTree) createGuildNode(n *tview.TreeNode, g discord.Guild) {
+func (gt *guildsTree) findUnreadChannelsAcrossAllGuilds() []discord.ChannelID {
+	if discordState == nil {
+		return nil
+	}
```
Redundant

marhaasa committed:
Remove redundant state check, add comments

marhaasa committed:
Fix: unread next/prev navigation, added API ack

ayn2op committed:
Merge branch 'main' into feature/add-unread-flags

marhaasa committed:
Merge branch 'main' into feature/add-unread-flags

ayn2op committed:
Merge branch 'main' into feature/add-unread-flags
